File: android/app/src/main/java/com/bensbible/app/viewmodel/NotesViewModel.kt
package com.bensbible.app.viewmodel

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.bensbible.app.data.AnnotationRepository
import com.bensbible.app.data.VerseAnnotationEntity
import com.bensbible.app.model.VerseID
import kotlinx.coroutines.launch

class NotesViewModel(
    private val repository: AnnotationRepository
) : ViewModel() {

    var notes by mutableStateOf<List<VerseAnnotationEntity>>(emptyList())
        private set

    var isEditorPresented by mutableStateOf(false)
    var editingAnnotation by mutableStateOf<VerseAnnotationEntity?>(null)
        private set
    var editingText by mutableStateOf("")

    fun load() {
        viewModelScope.launch {
            notes = repository.fetchAllNotes()
        }
    }

    fun beginEditing(annotation: VerseAnnotationEntity) {
        editingAnnotation = annotation
        editingText = annotation.noteText ?: ""
        isEditorPresented = true
    }

    fun saveEdit() {
        val annotation = editingAnnotation ?: return
        val verseID = VerseID(
            book = annotation.bookName,
            chapter = annotation.chapterNumber,
            verse = annotation.verseNumber
        )
        val text = editingText
        viewModelScope.launch {
            repository.saveNote(text, verseID)
            editingAnnotation = null
            editingText = ""
            load()
        }
    }

    fun cancelEdit() {
        editingAnnotation = null
        editingText = ""
    }

    fun deleteNote(annotation: VerseAnnotationEntity) {
        viewModelScope.launch {
            repository.deleteNote(annotation)
            load()
        }
    }
}

File: BensBible2/ViewModels/NotesViewModel.swift
import Foundation

@Observable
final class NotesViewModel {
    private var annotationService: AnnotationService?

    var notes: [VerseAnnotation] = []
    var isEditorPresented = false
    var editingAnnotation: VerseAnnotation?
    var editingText: String = ""

    func configure(annotationService: AnnotationService) {
        self.annotationService = annotationService
    }

    func load() {
        guard let service = annotationService else { return }
        notes = service.fetchAllNotes()
    }

    func beginEditing(_ annotation: VerseAnnotation) {
        editingAnnotation = annotation
        editingText = annotation.noteText ?? ""
        isEditorPresented = true
    }

    func saveEdit() {
        guard let annotation = editingAnnotation else { return }
        let verseID = VerseID(book: annotation.bookName, chapter: annotation.chapterNumber, verse: annotation.verseNumber)
        annotationService?.saveNote(text: editingText, for: verseID)
        editingAnnotation = nil
        editingText = ""
        load()
    }

    func cancelEdit() {
        editingAnnotation = nil
        editingText = ""
    }

    func deleteNote(_ annotation: VerseAnnotation) {
        let verseID = VerseID(book: annotation.bookName, chapter: annotation.chapterNumber, verse: annotation.verseNumber)
        annotationService?.saveNote(text: "", for: verseID)
        load()
    }
}
